### PR TITLE
Add a `query_paths` macro

### PR DESCRIPTION
**Pull request by shs96c**
This allows a genrule to find paths from source rules.

### CODE REVIEW

**Comment by shs96c** [comment]
Closes https://github.com/facebook/buck/issues/1520

File: src/com/facebook/buck/rules/BUCK [hunk 1]
```diff
@@ -47,6 +47,11 @@ java_immutables_library(
     ],
 )
 
+genrule(name = "foo",
```
Comment (ttsugriy) [inline src/com/facebook/buck/rules/BUCK]: changes to this build file should be part of an integration test - not production code ;)

File: src/com/facebook/buck/rules/macros/AbstractQueryPathsMacro.java [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+ * under the License.
+ */
+
+package com.facebook.buck.rules.macros;
+
+import com.facebook.buck.util.immutables.BuckStyleTuple;
+import org.immutables.value.Value;
+
+/**
+ * Class providing the type for macros used in `$(query_outputs ...)` macro strings. The
```
Comment (ttsugriy) [inline src/com/facebook/buck/rules/macros/AbstractQueryPathsMacro.java]: did you mean `$(query_paths ...)`?

**Comment by ttsugriy** [comment]
also added @asp2insp to take a look, since he's our query wizard :)

File: src/com/facebook/buck/rules/BUCK (hunk 1 above)
Comment (asp2insp) [reply src/com/facebook/buck/rules/BUCK]: If you want an existing test to model, here's where most of the genrule query tests live so far: https://github.com/facebook/buck/blob/master/test/com/facebook/buck/shell/GenruleDescriptionIntegrationTest.java and here's where their buck files are: https://github.com/facebook/buck/blob/master/test/com/facebook/buck/shell/testdata/genrule_classpath_filtering/BUCK.fixture

File: src/com/facebook/buck/rules/macros/QueryPathsMacroExpander.java [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+
+    return precomputedWork
+        .results
+        .stream()
+        .map(
+            queryTarget -> {
+              // What we do depends on the input.
+              if (QueryBuildTarget.class.isAssignableFrom(queryTarget.getClass())) {
+                BuildRule rule = resolver.getRule(((QueryBuildTarget) queryTarget).getBuildTarget());
+                return Optional.ofNullable(rule.getSourcePathToOutput())
```
Comment (asp2insp) [inline src/com/facebook/buck/rules/macros/QueryPathsMacroExpander.java]: What's the use case for the mixed-mode outputs/inputs here?

File: src/com/facebook/buck/rules/BUCK (hunk 1 above)
Comment (shs96c) [reply src/com/facebook/buck/rules/BUCK]: Ha! Ooops. :) Rolled back to master's

File: src/com/facebook/buck/rules/macros/AbstractQueryPathsMacro.java (hunk 2 above)
Comment (shs96c) [reply src/com/facebook/buck/rules/macros/AbstractQueryPathsMacro.java]: Yes. Fixed.

File: src/com/facebook/buck/rules/macros/QueryPathsMacroExpander.java (hunk 3 above)
Comment (shs96c) [reply src/com/facebook/buck/rules/macros/QueryPathsMacroExpander.java]: You want people to be able to use a query string that returns either rules or files and have it work in the same way, which makes it easier to iteratively futz with the query string as you tweak a build to get it Just Right.

Ideally, there would be just a single `query` macro that works no matter what it is you're trying to do. That's not where we are :)

**Comment by styurin** [comment]
Please, re-format Java files with google-java-format: https://github.com/facebook/buck/blob/master/CONTRIBUTING.md#code-style

**Pull request closed without merging**